Update CI Visibility documentation (#22273)
DataDog/documentation: content/en/tests/setup/java.md

## Extensions

The tracer exposes a set of APIs that can be used to extend its functionality programmatically.

### Adding custom tags to tests

To add custom tags include [opentracing-util][4] library as a compile-time dependency to your project.

> You can then add custom tags to your tests by using the current active span:

Reviewer: ```suggestion
You can then add custom tags to your tests by using the active span:
```

### Adding custom metrics to tests

> To add custom metrics include [opentracing-util][4] library as a compile-time dependency to your project.

Reviewer: ```suggestion
To add custom metrics, include the [`opentracing-util`][4] library as a compile-time dependency to your project.
```

If you are using a framework that is not supported, or an ad-hoc testing solution, you can harness the manual testing API, which also reports test results to the backend.

> To use manual testing API add [dd-trace-api][7] library as a compile-time dependency to your project.

Reviewer: ```suggestion
To use the manual testing API, add the [`dd-trace-api`][7] library as a compile-time dependency to your project.
```

Always call ``datadog.trace.api.civisibility.DDTestSession#end`` at the end so that all the test info is flushed to Datadog.

## Best practices

### Deterministic test parameters representation

Test Visibility works best when the [test parameters are deterministic][8] and stay the same between test runs.
If a test case has a parameter that varies between test executions (such as a current date, a random number, or an instance of a class whose `toString()` method is not overridden), some of the product features may not work as expected.
For example, the history of executions may not be available, or the test case may not be classified as flaky even if it exhibits flakiness.

The best way to fix this is to make sure that the test parameters are the same between test runs.

> In JUnit 5 this can also be addressed by [customizing the string representation of the test parameters][9] without changing their values.

Reviewer: ```suggestion
In JUnit 5, this can also be addressed by [customizing the string representation of the test parameters][9] without changing their values.
```

> To do this use `org.junit.jupiter.api.Named` interface or change the `name` parameter of the `org.junit.jupiter.params.ParameterizedTest` annotation:

Reviewer: ```suggestion
To do so, use `org.junit.jupiter.api.Named` interface or change the `name` parameter of the `org.junit.jupiter.params.ParameterizedTest` annotation:
```

   // In the argument provider method it is wrapped with Named to ensure it has a deterministic name.
}

static Stream<Arguments> namedArguments() {
    return Stream.of(
            Arguments.of(
                    "a string",
                    Named.of("current date", new Date())),
            Arguments.of(
                    "another string",
                    Named.of("a date in the future", new Date(System.currentTimeMillis() + TimeUnit.DAYS.toMillis(1))))
    );
}
```

```java
@ParameterizedTest(name = "[{index}] {0}, a random number from one to ten")
@MethodSource("randomArguments")
void anotherParameterizedTest(String s, int i) {
  // The second parameter in this test case is non-deterministic.
  // The name of the parameterized test is customized to ensure it has a deterministic name.
}

static Stream<Arguments> randomArguments() {
    return Stream.of(
            Arguments.of("a string", ThreadLocalRandom.current().nextInt(10) + 1),
            Arguments.of("another string", ThreadLocalRandom.current().nextInt(10) + 1)
    );
}
```

## Troubleshooting

### The tests are not appearing in Datadog after enabling CI Visibility in the tracer

> Verify that the tracer is injected into your build process: examine your build's logs.

Reviewer: ```suggestion
Verify that the tracer is injected into your build process by examining your build's logs.
```

> If the injection is successful, you will see a line containing `DATADOG TRACER CONFIGURATION`.

Reviewer: ```suggestion
If the injection is successful, you can see a line containing `DATADOG TRACER CONFIGURATION`.
```

Ensure that the `dd.civisibility.enabled` property (or `DD_CIVISIBILITY_ENABLED` environment variable) is set to `true` in the tracer arguments.

> Try running your build with tracer debug logging enabled (set `DD_TRACE_DEBUG` environment variable to `true`).

Reviewer: ```suggestion
Try running your build with tracer debug logging enabled by setting the `DD_TRACE_DEBUG` environment variable to `true`.
```

If the plugin interferes with the build, disable it by adding `dd.civisibility.compiler.plugin.auto.configuration.enabled=false` to the list of `-javaagent` arguments
(or by setting `DD_CIVISIBILITY_COMPILER_PLUGIN_AUTO_CONFIGURATION_ENABLED=false` environment variable).

### Builds fails because dd-javac-plugin-client artifact cannot be found

It is possible that the Java compiler plugin injected into the build is not available if the build uses a custom artifactory storage or if it is run in offline mode.

If this is the case, you can disable plugin injection by adding `dd.civisibility.compiler.plugin.auto.configuration.enabled=false` to the list of `-javaagent` arguments

> (or by setting `DD_CIVISIBILITY_COMPILER_PLUGIN_AUTO_CONFIGURATION_ENABLED=false` environment variable).

Reviewer: ```suggestion
(or by setting the `DD_CIVISIBILITY_COMPILER_PLUGIN_AUTO_CONFIGURATION_ENABLED` environment variable to false).
```
